### PR TITLE
SDL-0253 Check UnsubscribeVehicleData RPC processing with StabilityControlsStatus

### PR DESCRIPTION
**Pull request by LitvinenkoIra**
ATF Test Scripts to check [0253-New-vehicle-data-StabilityControlsStatus](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0253-New-vehicle-data-StabilityControlsStatus.md)

[Jira Task](https://adc.luxoft.com/jira/browse/FORDTCN-6007)

This PR is **ready** for review.

### Summary
Check UnsubscribeVehicleData RPC with:
- `stabilityControlsStatus` and other parameters

- `stabilityControlsStatus` parameter only which is NOT allowed by Policies

- `stabilityControlsStatus` parameter which is NOT allowed by Policies and other parameters which is allowed by Policies

- `stabilityControlsStatus` parameter only, invalid response from HMI related to this parameter

- `stabilityControlsStatus` and other parameters, invalid response from HMI related to `stabilityControlsStatus` parameter

- `stabilityControlsStatus` and other parameters, invalid response from HMI related to other parameter

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/API/VehicleData/commonVehicleData.lua [hunk 1]
```diff
@@ -321,23 +351,72 @@ function m.processRPCSubscriptionSuccess(pRpcName, pData)
...
     end)
   local mobResParams = m.cloneTable(hmiResParams)
-  mobResParams.success = true
-  mobResParams.resultCode = "SUCCESS"
+  mobResParams = {
+    [pDisallowedData] = {
+      resultCode = "DISALLOWED",
+      dataType = m.allVehicleData[pDisallowedData].type
+    },
+    ["success"] = true,
```
Comment (ydementieiev) [inline test_scripts/API/VehicleData/commonVehicleData.lua:400]: @LitvinenkoIra why not `success = true` ?

File: test_scripts/API/VehicleData/commonVehicleData.lua [hunk 2]
```diff
@@ -321,23 +351,72 @@ function m.processRPCSubscriptionSuccess(pRpcName, pData)
...
+    ["resultCode"] = "SUCCESS",
+    ["info"] = "'stabilityControlsStatus' disallowed by policies."
+  }
   m.getMobileSession():ExpectResponse(cid, mobResParams)
 end
 
+function m.processRPCSubscriptionGenericError(pRpcName, pReqParams, pHmiResParams)
+  local reqParams = {}
+  for k,v in pairs(pReqParams) do
```
Comment (ydementieiev) [inline test_scripts/API/VehicleData/commonVehicleData.lua:409]: @LitvinenkoIra  unused k
```suggestion
  for _,v in pairs(pReqParams) do
```

File: test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+common.Step("Start SDL, init HMI, connect default mobile", common.start)
+common.Step("Register App", common.registerApp)
+common.Step("PTU", common.policyTableUpdate, { common.ptUpdate })
+common.Step("Activate App", common.activateApp)
+common.Step("Subscribe on gps, speed, rpm, stabilityControlsStatus", 
+  common.processRPCSubscriptionSuccess, {"SubscribeVehicleData", vehicle_data_items })
+
+common.Title("Test")
+common.Step("Unsubscribe from gps, speed, rpm, stabilityControlsStatus",
+  common.processRPCSubscriptionSuccess, {"UnsubscribeVehicleData", vehicle_data_items })
```
Comment (VjKlepikov) [inline test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua:48]:
```suggestion
  common.processRPCSubscriptionSuccess, { "UnsubscribeVehicleData", vehicle_data_items })
```
Please add whitespaces here, to improve readability. Please check all files of this PR

File: test_scripts/API/VehicleData/commonVehicleData.lua [hunk 4]
```diff
@@ -321,23 +351,72 @@ function m.processRPCSubscriptionSuccess(pRpcName, pData)
...
-  m.getHMIConnection():ExpectRequest("VehicleInfo." .. pRpcName, reqParams)
+  local cid = m.getMobileSession():SendRPC(pRpcName, mobReqParams)
+  m.getHMIConnection():ExpectRequest("VehicleInfo." .. pRpcName, hmiReqParams)
   :Do(function(_, data)
       m.getHMIConnection():SendResponse(data.id, data.method, "SUCCESS", hmiResParams)
     end)
   local mobResParams = m.cloneTable(hmiResParams)
-  mobResParams.success = true
-  mobResParams.resultCode = "SUCCESS"
+  mobResParams = {
```
Comment (VjKlepikov) [inline test_scripts/API/VehicleData/commonVehicleData.lua:395]: mobResParams is overwritten.  Please add AllowedData structure to mobResParams

File: test_scripts/API/VehicleData/commonVehicleData.lua (hunk 1 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/commonVehicleData.lua]: @ydementieiev Done in 69c60b2146aef154480b0ad4d6d4dcfa593a54df

File: test_scripts/API/VehicleData/commonVehicleData.lua (hunk 2 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/commonVehicleData.lua]: @ydementieiev Done in 69c60b2146aef154480b0ad4d6d4dcfa593a54df

File: test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua (hunk 3 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua]: @VjKlepikov Done in 69c60b2146aef154480b0ad4d6d4dcfa593a54df

File: test_scripts/API/VehicleData/commonVehicleData.lua (hunk 4 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/commonVehicleData.lua]: @VjKlepikov Done in 69c60b2146aef154480b0ad4d6d4dcfa593a54df

File: test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+--
+-- Precondition:
+-- 1) SDL and HMI are started
+-- 2) App is registered
+-- 3) PTU is successfully performed
+-- 4) App is activated
+-- 5) App is subscribed on StabilityControlsStatus and some other vehicle data
+--
+-- Steps:
+-- 1) App send UnsubscribeVehicleData requests to SDL (for all subscribed vehicle data items)
```
Comment (AByzhynar) [inline test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua:14]: @LitvinenkoIra sendS

File: test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua [hunk 6]
```diff
@@ -0,0 +1,36 @@
+---------------------------------------------------------------------------------------------------
+-- Proposal: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0253-New-vehicle-data-StabilityControlsStatus.md
+--
+-- Description: UnsubscribeVehicleData RPC with `stabilityControlsStatus` parameter only is NOT allowed by Policies
+--
+-- Precondition:
+-- 1) SDL and HMI are started
+-- 2) App is registered
+-- 3) PTU is successfully performed
```
Comment (AByzhynar) [inline test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua:9]: @LitvinenkoIra Add comment please what is performed in PTU (minimal) here or just before `common.ptUpdateMin`

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua [hunk 7]
```diff
@@ -0,0 +1,44 @@
+---------------------------------------------------------------------------------------------------
+-- Proposal: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0253-New-vehicle-data-StabilityControlsStatus.md
+--
+-- Description: UnsubscribeVehicleData RPC with `stabilityControlsStatus` parameter 
+-- which is NOT allowed by Policies and other parameters which is allowed by Policies
```
Comment (AByzhynar) [inline test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua:5]: @LitvinenkoIra parameters which are allowed

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+-- 1) SDL and HMI are started
+-- 2) App is registered
+-- 3) PTU is successfully performed
+-- 4) App is activated
+-- 5) App is subscribed on gps vehicle data
+--
+-- Steps:
+-- 1) App send UnsubscribeVehicleData (with stabilityControlsStatus = true and gps = true) request to SDL
+-- SDL does:
+--  - cut stabilityControlsStatus off
```
Comment (AByzhynar) [inline test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua:17]: @LitvinenkoIra cuts

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+--  - send VehicleInfo.UnsubscribeVehicleData request with (gps = true) to HMI
```
Comment (AByzhynar) [inline test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua:18]: @LitvinenkoIra sends

File: test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua [hunk 10]
```diff
@@ -0,0 +1,36 @@
+---------------------------------------------------------------------------------------------------
+-- Proposal: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0253-New-vehicle-data-StabilityControlsStatus.md
+--
+-- Description: UnsubscribeVehicleData RPC with `stabilityControlsStatus` parameter only is NOT allowed by Policies
```
Comment (AByzhynar) [inline test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua:4]: @LitvinenkoIra Unclear description. 
- UnsubscribeVehicleData RPC with the only parameter `stabilityControlsStatus`, which is NOT allowed by Policies

File: test_scripts/API/VehicleData/StabilityControlsStatus/021_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI.lua [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+--
+-- Precondition:
+-- 1) SDL and HMI are started
+-- 2) App is registered
+-- 3) PTU is successfully performed
+-- 4) App is activated
+-- 5) App is subscribed on StabilityControlsStatus vehicle data
+--
+-- Steps:
+-- 1) App send UnsubscribeVehicleData (with stabilityControlsStatus = true) request to SDL
```
Comment (AByzhynar) [inline test_scripts/API/VehicleData/StabilityControlsStatus/021_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI.lua:15]: @LitvinenkoIra sends

File: test_scripts/API/VehicleData/commonVehicleData.lua (hunk 1 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/commonVehicleData.lua]: @ydementieiev Done in 69c60b2

File: test_scripts/API/VehicleData/commonVehicleData.lua (hunk 2 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/commonVehicleData.lua]: @ydementieiev Done in 69c60b2

File: test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua (hunk 3 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua]: @ydementieiev Done in 69c60b2

File: test_scripts/API/VehicleData/commonVehicleData.lua (hunk 4 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/commonVehicleData.lua]: @ydementieiev Done in 69c60b2

File: test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua (hunk 5 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/018_UnsubscribeVehicleData_StabilityControlsStatus_and_other_parameters.lua]: @AByzhynar Done in 4689ec2

File: test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua (hunk 6 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua]: @AByzhynar Done in 4689ec2

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua (hunk 7 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua]: @AByzhynar Done in 4689ec2

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua (hunk 8 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua]: @AByzhynar Done in 4689ec2

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua (hunk 9 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua]: @AByzhynar Done in 4689ec2

File: test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua (hunk 10 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua]: @AByzhynar Done in 4689ec2

File: test_scripts/API/VehicleData/StabilityControlsStatus/021_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI.lua (hunk 11 above)
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/021_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI.lua]: @AByzhynar Done in 4689ec2

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua (hunk 9 above)
Comment (AByzhynar) [reply test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua]: @LitvinenkoIra This is wrong. Please get the idea from this comment: https://github.com/LuxoftSDL/sdl_atf_test_scripts/pull/15/files#r425053019

File: test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+---------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/VehicleData/commonVehicleData')
+
+--[[ Scenario ]]
+common.Title("Preconditions")
+common.Step("Clean environment", common.preconditions)
+common.Step("Start SDL, init HMI, connect default mobile", common.start)
+common.Step("Register App", common.registerApp)
+common.Step("PTU", common.policyTableUpdate, { common.ptUpdateMin })
```
Comment (aderiabin) [inline test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua:28]: @LitvinenkoIra  Performed PTU disallows stabilityControlsStatus parameter not only for tested in this script RPC (SubscribeVehicleData). So the script can miss possible class of defects when policy works incorrect for new parameter.
I would propose to disable only new parameter for only tested RPC (stabilityControlsStatus parameter for SubscribeVehicleData)

This comment relates next script too.

File: test_scripts/API/VehicleData/StabilityControlsStatus/021_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI.lua [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+--
+-- Precondition:
+-- 1) SDL and HMI are started
+-- 2) App is registered
+-- 3) PTU is successfully performed
+-- 4) App is activated
+-- 5) App is subscribed on StabilityControlsStatus vehicle data
+--
+-- Steps:
+-- 1) App sends UnsubscribeVehicleData (with stabilityControlsStatus = true) request to SDL
```
Comment (aderiabin) [inline test_scripts/API/VehicleData/StabilityControlsStatus/021_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI.lua:15]: @LitvinenkoIra  Please describe each checked by the script case explicitly here

This comment relates to next scripts too

File: test_scripts/API/VehicleData/StabilityControlsStatus/023_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI_to_other_params.lua [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+-- 1) App sends UnsubscribeVehicleData (with stabilityControlsStatus = true and gps = true) request to SDL
+-- SDL does:
+--  - sends VehicleInfo.UnsubscribeVehicleData (with stabilityControlsStatus = true and gps = true) request to HMI
+-- HMI sends invalid VehicleInfo.UnsubscribeVehicleData response related to `gps` parameter
+-- SDL does:
+--  - send UnsubscribeVehicleData response with (success: false, resultCode: "GENERIC_ERROR")
+-- 2) HMI sends VehicleInfo.OnVehicleData notifications with StabilityControlsStatus data
+-- SDL does:
+--  - send OnVehicleData notification with received from HMI data to App
+-- 2) HMI sends VehicleInfo.OnVehicleData notifications with GPS data
```
Comment (aderiabin) [inline test_scripts/API/VehicleData/StabilityControlsStatus/023_UnsubscribeVehicleData_StabilityControlsStatus_invalid_response_from_HMI_to_other_params.lua:24]: @LitvinenkoIra 
```suggestion
-- HMI sends VehicleInfo.OnVehicleData notifications with GPS data
```

**Comment by LitvinenkoIra** [comment]
@aderiabin all comments are processed in e56e022ba2599531f0e50f6698195dee436c8543

File: test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+--      stabilityControlsStatus = { resultCode = "DISALLOWED", dataType = "VEHICLEDATA_STABILITYCONTROLSSTATUS"})
+---------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/VehicleData/commonVehicleData')
+
+--[[ Local Functions ]]
+function ptUpdate(pTbl)
+  pTbl.policy_table.app_policies[common.getConfigAppParams().fullAppID].groups = { "Base-4", "Emergency-1" }
+  local grp = pTbl.policy_table.functional_groupings["Emergency-1"]
+  grp.rpcs.SubscribeVehicleData.parameters = {
```
[the remark this answers is not in the document]
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/020_UnsubscribeVehicleData_StabilityControlsStatus_partially_disallowed_by_policy.lua:31]: @aderiabin please check https://github.com/LuxoftSDL/sdl_atf_test_scripts/pull/16/files#diff-b791918424d6eff05eb0d1bc9dde4428R34

File: test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+--      stabilityControlsStatus = {resultCode = "DISALLOWED", dataType = "VEHICLEDATA_STABILITYCONTROLSSTATUS"}) to App
+---------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/VehicleData/commonVehicleData')
+
+--[[ Local Functions ]]
+function ptUpdate(pTbl)
+  pTbl.policy_table.app_policies[common.getConfigAppParams().fullAppID].groups = { "Base-4", "Emergency-1" }
+  local grp = pTbl.policy_table.functional_groupings["Emergency-1"]
+  grp.rpcs.SubscribeVehicleData.parameters = common.EMPTY_ARRAY
```
[the remark this answers is not in the document]
Comment (LitvinenkoIra) [reply test_scripts/API/VehicleData/StabilityControlsStatus/019_UnsubscribeVehicleData_StabilityControlsStatus_disallowed_by_policy.lua]: @aderiabin please check 57ee7e3

**Pull request merged**